ArduPlane/systemid.cpp

```diff
@@ -0,0 +1,243 @@
+    AP_GROUPINFO_FLAGS("_AXIS", 1, AP_SystemID, axis, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: _MAGNITUDE
+    // @DisplayName: System identification Chirp Magnitude
+    // @Description: Magnitude of sweep in deg, deg/s and 0-1 for mixer outputs.
+    // @User: Standard
+    AP_GROUPINFO("_MAGNITUDE", 2, AP_SystemID, waveform_magnitude, 5),
+
+    // @Param: _F_START_HZ
+    // @DisplayName: System identification Start Frequency
+    // @Description: Frequency at the start of the sweep
+    // @Range: 0.01 100
+    // @Units: Hz
+    // @User: Standard
+    AP_GROUPINFO("_F_START_HZ", 3, AP_SystemID, frequency_start, 0.5f),
+
+    // @Param: _F_STOP_HZ
+    // @DisplayName: System identification Stop Frequency
+    // @Description: Frequency at the end of the sweep
+    // @Range: 0.01 100
+    // @Units: Hz
+    // @User: Standard
+    AP_GROUPINFO("_F_STOP_HZ", 4, AP_SystemID, frequency_stop, 15),
+
+    // @Param: _T_FADE_IN
+    // @DisplayName: System identification Fade in time
+    // @Description: Time to reach maximum amplitude of sweep
+    // @Range: 0 20
+    // @Units: s
+    // @User: Standard
+    AP_GROUPINFO("_T_FADE_IN", 5, AP_SystemID, time_fade_in, 5),
+
+    // @Param: _T_REC
+    // @DisplayName: System identification Total Sweep length
+    // @Description: Time taken to complete the sweep
+    // @Range: 0 255
+    // @Units: s
+    // @User: Standard
+    AP_GROUPINFO("_T_REC", 6, AP_SystemID, time_record, 70),
+
+    // @Param: _T_FADE_OUT
+    // @DisplayName: System identification Fade out time
+    // @Description: Time to reach zero amplitude at the end of the sweep
+    // @Range: 0 5
+    // @Units: s
+    // @User: Standard
+    AP_GROUPINFO("_T_FADE_OUT", 7, AP_SystemID, time_fade_out, 1),
+
+    // @Param: _XY_CTRL_MUL
+    // @DisplayName: System identification XY control multiplier
+    // @Description: A multiplier for the XY velocity and position controller when using systemID in VTOL modes that do horizontal position and velocity control
+    // @Range: 0.1 1.0
+    // @User: Standard
+    AP_GROUPINFO("_XY_CTRL_MUL", 8, AP_SystemID, xy_control_mul, 0.25),
+    
+    AP_GROUPEND
+};
+
+AP_SystemID::AP_SystemID(void)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+/*
+  start systemid
+ */
+void AP_SystemID::start()
+{
+    start_axis = axis;
+
+    // check if enabled
+    if (start_axis == AxisType::NONE) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "SystemID: No axis selected");
+        return;
+    }
+    if (!plane.control_mode->supports_systemid()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "SystemID: Not supported in mode %s", plane.control_mode->name());
+        return;
+    }
+    if (!hal.util->get_soft_armed()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "SystemID: must be armed");
+        return;
+    }
+    if (!plane.quadplane.enabled()) {
```

Comment:
This case should be handled by `plane.control_mode->supports_systemid()`

Reply:
bump

Reply:
The message is still useful I guess. But it need to be above the mode check or it will never be seen.

Reply:
moved above mode check